### ISSUE TITLE
WSL2: Ubuntu 18.04 Serial devices are no longer mounted properly

### ISSUE THREAD

**Issue by strike5150** [microsoft/WSL#4322]
* Your Windows build number:  Microsoft Windows [Version 10.0.18932.1000]

* What you're doing and what's happening:  
- run WSL2 Ubuntu 18.04
- plugin USB to serial device
- watch it mount under COM devices in device manager
- attempt to open the device in WSL2
- cat /dev/ttyS[X]

* What's wrong / what should be happening instead:
This works under WSL1 flawlessy, it no longer works in WSL2
in WSL2 we can clearly see it is not connected or configured
/dev/ttyS3, UART: unknown, Port: 0x02e8, IRQ: 3
![COM3-wsl-pic](https://user-images.githubusercontent.com/1725627/61539227-6fe5ae00-aa11-11e9-87c8-aaff4fabf2ed.png)

![COM3-pic](https://user-images.githubusercontent.com/1725627/61539193-5a708400-aa11-11e9-8331-9640b5ce35bb.png)

**Comment by rubberduck203** [comment]
It’s clearly called out in the [FAQ](https://docs.microsoft.com/en-us/windows/wsl/wsl2-faq#can-i-access-the-gpu-in-wsl-2-are-there-plans-to-increase-hardware-support) that there is currently no support for USB in WSL2. 

I’d like to toss my voice in here that this should be a priority feature for WSL2. 

One of the big benefits of WSL2 is the opportunity to run Docker natively instead of on Docker for Windows. [It is notoriously hard to get access to USB devices in Docker on non-Linux systems.](https://github.com/docker/for-mac/issues/900)

I commonly use Docker for creating tool-chain images, but this falls apart when you’re working with embedded devices, where you need to connect to a USB device or serial port in order to flash/debug on device. If I can run a docker container natively *and* access USB ports with WSL2, I can make an *extremely* strong argument for switching from Mac to Windows.

**Comment by linweifu** [comment]
up

**Comment by zawawimanja** [comment]
so that why I cannot see my usb device in ubuntu.hurmmm

**Comment by rubberduck203** [comment]
It would be really nice if someone could chime in letting us know if this is even technically feasible.

**Comment by sebbean** [comment]
would love to know

**Comment by chenhu66** [comment]
what a pity! i gotta roll back to wsl 1. Hopefully, usb support comes up soom for wsl2

**Comment by tleroy-dev** [comment]
> 
> 
> what a pity! i gotta roll back to wsl 1. Hopefully, usb support comes up soom for wsl2

Can I have the deatiled step to roll back from WSL2 to WSL1 ?
I found under https://docs.microsoft.com/en-us/windows/wsl/release-notes#build-18917 that WSL2 was introduced at build version 18917. Now, how can I re-install WSL from a specific build number ? Thank you.

**Comment by rubberduck203** [comment]
I thought you could have WSL1 and WSL2 installed side by side?

**Comment by tleroy-dev** [comment]
@rubberduck203 it's not clear to me how to check which WSL version we are running. It is not described in the app store. From [online information i found](https://medium.com/@gmusumeci/linux-on-windows-totally-how-to-install-wsl-1-and-wsl-2-307c9dd38a36):

> We need to run Windows Build 16215 or later to install WSL 1.
> 
> WSL 2 is only available in Windows Builds 18917 or higher.

**Comment by bburky** [comment]
@tleroy-dev If your systems supports WSL2, you can choose WSL1 or WSL2 per distro.

`VERSION`  from `wsl --list` shows if you're using WSL1 or WSL 2:
```
> wsl --list --verbose
  NAME                   STATE           VERSION
* docker-desktop-data    Stopped         2
  docker-desktop         Stopped         2
  Ubuntu-18.04           Stopped         2
  Fedora-32              Stopped         2
  Ubuntu                 Stopped         1
```

The normal recommendation is to use WSL2 for installation of all new WSL distros. Note, this doesn't affect your current installed WSL distros, if you have any already they will stay on WSL1.
```
wsl --set-default-version 2
```

If you want to manually upgrade a distro to WSL2 or "downgrade" it to WSL1, you can use `--set-version`. I keep a second WSL1 distro around to have functioning `/dev/ttyS*` devices.
```
wsl --set-version Ubuntu 1
```

**Comment by tleroy-dev** [comment]
@bburky Thanks for your response.

```
PS C:\Users\nxf43719> wsl --list --verbose
/bin/bash: --: invalid option
```
[From that documentation](https://docs.microsoft.com/pl-pl/windows/wsl/wsl-config), it looks like the wsl.exe command works only on build higher than 1903, older version use wslconfig.exe. Result of equivalent : 
```
PS C:\Users\nxf43719> wslconfig.exe /list /all
Distributions du sous-système Windows pour Linux :
Ubuntu (par défaut)
```
It's in french in my case (system default), it does not output the version specifically, I guess from hat I described earlier, being on a pre-WSL2 build, only WSL 1 is enabled in my case.

**Comment by bburky** [comment]
@tleroy-dev Yes, sounds like you're definitely on WSL1 then.

**Comment by tleroy-dev** [comment]
I have opened a new ticket for my problem under WSL 1 (#5203)

**Comment by AsciiWolf** [comment]
Any news regarding USB devices support in WSL2?

**Comment by ditchfieldcaleb** [comment]
Just upgraded to WSL2 for the native Docker integration, but frustrated that I can't continue to do Arduino development natively in WSL anymore without switching back and forth between WSL versions... any updates here?

**Comment by Ludorg** [comment]
> 
> 
> Just upgraded to WSL2 for the native Docker integration, but frustrated that I can't continue to do Arduino development natively in WSL anymore without switching back and forth between WSL versions... any updates here?

Same frustation and same behavior here...

**Comment by GearWalker** [comment]
I'm also working with embedded system build environment, really need this too!

**Comment by kategray** [comment]
This is a huge blocker as well.  Using WSL1 as a workaround, but there needs to be some way to get USB->Serial working on WSL2.

**Comment by allen-chu** [comment]
Any updates on this? This is a showstopper for moving to WSL for embedded Linux development. There are issues with buildroot + WSL1 so that's not an option either.

**Comment by mkanet** [comment]
I guess I can forget about installing Home Assistant under WSL2 anytime soon.  My [HUSBZB-1-USB-HUB](https://www.amazon.com/GoControl-CECOMINOD016164-HUSBZB-1-USB-Hub/dp/B01GJ826F8) won't mount.

**Comment by cranphin** [comment]
Another vote for this, being able to attach serial over usb embedded devices to my docker containers would be a game changer

**Comment by OmNamasi** [comment]
Bummer! I was happy to setup WSL 2 & install Ubuntu to build my project on linux. However i couldn't flash or access the serial/COM USB port from WSL 2. This feature is a huge blocker. Can you pull-in the device access support from WSL 1 immediately?

**Comment by dattrax** [comment]
I could really do with this to flash an embedded serial device.

**Comment by mastrolinux** [comment]
Arduino CIO here, as Arduino, we would like to contribute in making this happen. Allowing developers to work with WSL and Arduino boards would be fanstastic. As of now, our arduino-cli inside WSL does not allow people to see the list of the boards. Mainly related to https://github.com/microsoft/WSL/issues/3099 and the closed one: https://github.com/microsoft/WSL/issues/1929. Thank you for your great work!

**Comment by TaDahCorp** [comment]
Another vote for USB-serial support, WSL2 was doing so well otherwise on first try...

**Comment by cduplantis** [comment]
I wanted to build and burn an custom raspberry image to a flash drive from a WSL2  session. My first test (writing the Raspberry Pi OS) using the `dd` utility, I found out that writing to host USB ports is not possible.

**Comment by pianocomposer321** [comment]
I have a workaround for Arduino at least:

I downloaded the arduino-cli and put the exe in the Program Files (x86)/Arduino folder, then added that folder to path. Then you can access the exe's from WSL (see https://docs.microsoft.com/en-us/windows/wsl/interop). You can run the command just as you would from windows. The only difference is that instead of just `arduino-cli` you have to append .exe, as in `arduino-cli.exe`.

So to create a new arduino project, I do the following:

```
cd ~/arduino-projects
arduino-cli.exe sketch new cli-test && cd cli-test
```
Then edit the .ino file with neovim:
```
nvim cli-test.ino
```
And finally compile and upload the sketch:
```
arduino-cli.exe compile --fqbn arduino:avr:mega ~/arduino-projects/cli-test
arduino-cli.exe upload -p COM4 --fqbn arduino:avr:mega ~/arduino-projects/cli-test
```

Note that using full path (~/arduino-projects/cli-test) rather than just . is important. If you just use . , (for reasons I don't understand), you will get an error:
```
avrdude: can't open input file build\arduino.avr.mega/..ino.hex: No such file or directory
avrdude: read from file 'build\arduino.avr.mega/..ino.hex' failed
Error during Upload: uploading error: uploading error: exit status 1
```

Also note that the last two commands could be run from inside (neo)vim with :!, in which case the command would look like this:

```
:!arduino-cli.exe compile --fqbn arduino:avr:mega %:p:h
:!arduino-cli.exe upload -p COM4 --fqbn arduino:avr:mega %:p:h
```

You could also add a mapping in your vimrc (or init.vim) for these commands, or add a makefile.

HTH,
pianocomposer321

**Comment by badcf00d** [comment]
This is very much a workaround, but I thought I'd share the what I've done for ESP32 development until WSL 2 has serial access in case it's useful to anyone. You could probably follow a similar approach for other microcontrollers.

The WSL 2 virtual drives are mounted as network drives in Windows at `\\wsl$`, so you can do your builds in WSL, and then open a powershell window in the WSL directory with your build. You can then write some powershell scripts to flash and monitor over your COM ports, for the esp32 this works fairly well thanks to the python scripts they provide from [esptool](https://pypi.org/project/esptool/) and [idf_monitor.py](https://github.com/espressif/esp-idf/blob/master/tools/idf_monitor.py)

**flash.ps1**
```
# Make sure you have python, and esptool installed (pip install esptool)
# Usage: flash.ps1 COM6 

esptool.py -b 1152000 -p $args[0] write_flash 0x1000 .\build\bootloader\bootloader.bin 0x8000 .\build\partitions_two_ota.bin 0x10000 .\build\esp32-binary.bin
```

**monitor.ps1**
```
# Usage: monitor.ps1 COM6 

python idf_monitor.py --baud 921600 --port $args[0] .\build\esp32-binary.elf
```

**Comment by pianocomposer321** [comment]
@badcf00d Once the PowerShell scripts are written, you could probably even just run them from the wsl terminal, similarly to how you can run arduino-cli.exe from wsl.

**Comment by badcf00d** [comment]
> @badcf00d Once the PowerShell scripts are written, you could probably even just run them from the wsl terminal, similarly to how you can run arduino-cli.exe from wsl.

Oh yeh, good point actually, that's much easier: `powershell.exe ./monitor.ps1 COM6`.

**Comment by pianocomposer321** [comment]
Or even put all the scripts in the same folder and add them to PATH (in windows, not wsl (I think)). That way you don't have to be in the same directory as the scripts when you run it.

**Comment by caowent** [comment]
A better serial debugging tool is [sisterm](https://crates.io/crates/sisterm)
1. Command line form
2. It is a native exe application built using rust
3. It supports custom syntax highlighting
4. It works well in windows terminal
5. If it is only for debugging the serial port, such as embedded development, it consumes less memory than wsl1.
6. The simple start command in powershell: sist -l com3 -s 1500000

**Comment by Matheus-Garbelini** [comment]
any updates on this?

**Comment by jimmyw** [comment]
Please have a look at this soon!

**Comment by imadjamil** [comment]
This is really a showstopper for WSL and should be given more attention!

**Comment by rubendibattista** [comment]
> 
> 
> @rhyek I'm sure you've considered this already, but why not install Linux on your Windows laptop? With the right desktop environment, you can have just as good window management and just as good (or even identical) keyboard shortcuts as OSX.

Well, Linux on a Windows laptop is a suboptimal solution at the very least IMHO. 

* Battery Life: In my experience, a Laptop with generic drivers on Linux performs 30% worse than the same exact laptop with Windows installed.

* General compatibility. As an example I would bring the compatibility with [Clickshare](https://www.barco.com/en/clickshare), to make presentations. Lots of corporate environments have it, but you can't use on Linux. ANother example is the Touchpad drivers that are very often extremely poor on Linux, making the experience extremely frustrating.

* Gaming: It's very annoying to reboot on a system every time you want to play a game. 

WSL allows to work on Windows, having all the benefits of Windows world, but being able to switch on a fast Unix environment with a click. 

At the contrary of @rhyek, for me the current WSL experience is good enough to legitimate the switch from macOS to Windows. That said, it would be great to have Serial Device compatibility back again.

**Comment by pianocomposer321** [comment]
@rubendibattista Those are some very interesting points. Not trying to start an argument, so please don't take this that way:

First of all, you mentioned battery life. I don't have that much experience with Linux when my laptop wasn't plugged in, but I never had any issues. I'm sure it differs from laptop to laptop, battery to battery, and likely distro to distro.

Next, when it comes to compatibility:
There's a saying that goes "Anything that windows can do, Linux can do for free". Whatever program you are used to that doesn't run on Linux, there is a 99% chance that there is a free alternative that will. However, if you are a part of a team, and all of your teammates are running Windows machines, I can see why you might want to just keep it consistent. However, if you are "flying solo", this really isn't an issue.

For games, there are a number of factors involved here. First of all, gaming on Linux has gotten much, much better the last year or so, and there are distros (like Manjaro) made specifically for gaming. If you want something a little more versatile, Pop OS is a great choice too. It's based on Ubuntu and should provide a great gaming and programming experience. Also, if @rhyek isn't going to be programming on his windows machine anyway, then booting it up to play games really isn't that different than rebooting the same one. The gaming experience on OSX is just as bad (arguably worse) than gaming on Linux.

**Comment by cascornelissen** [comment]
@rhyek, @pianocomposer321, @rubendibattista; could you guys move this discussion somewhere else? It's only slightly related to the original issue and I'm subscribed to email updates (just like 30+ other people) because I'm interested in when this gets fixed, not in the millionth OS discussion.

**Comment by pianocomposer321** [comment]
@cascornelissen Not sure where we would move it to, but that's ok, it's not really that important, and I apologize for getting the discussion so off track ;-). I just wanted to make sure @rhyek had considered the other options.

**Comment by mirchat** [comment]
Are the serial devices mounted OK using Ubuntu 20 on WSL2 ?

**Comment by mastrolinux** [comment]
No, the issue is still present on Ubuntu 20. Not sure if we need to open another bug about it. @mirchat

**Comment by Matheus-Garbelini** [comment]
This is really sad as the whole Arduino, Platformio, Mbed, and firmware developers benefit a lot from serial port over WSL2.

**Comment by f4grx** [comment]
Please make serial devices work in wsl2, this is a showstopper for embedded development on windows.

**Comment by afilip1** [comment]
I would also like to express support for this feature. As an embedded developer, it's currently preventing me from using WSL2 in my workflow.

**Comment by abobija** [comment]
Any updates on this?
My ESP32 impatiently waiting to be flashed directly from WSL2 :sweat_smile:

**Comment by f4grx** [comment]
@abobija meanwhile you can use MSYS2 / MINGW64 with python installed to flash the esp32. just replace /mnt/drive by /drive and the --port by a real windows com port.

**Comment by abobija** [comment]
Thank you @f4grx I know, I already flash it like that, but I'm looking forward to flash it directly from wsl2, without any improvisations

**Comment by f4grx** [comment]
Same here, of course! it would be much easier. comm ports are a basic OS feature that should work from wsl.

**Comment by Matheus-Garbelini** [comment]
Indeed, Serial Port on WSL2 is a dream of every firmware engineer. Hope Microsoft can figure out how to add support for that soon.

**Comment by vladkosarev** [comment]
Still using wsl1 because of this and faster host-shared file system.

**Comment by abobija** [comment]
For those who wait WSL2 USB support for flashing [ESP-IDF](https://github.com/espressif/esp-idf) applications, I've wrote [this Shell script](https://github.com/abobija/idfx) with the help of which you can flash project with command `idfx flash COM3` from WSL2. Or for monitoring: `idfx monitor COM3`.

And [here](https://gist.github.com/abobija/2f11d1b2c7cb079bec4df6e2348d969f) is full tutorial how to install ESP-IDF on WSL2 (Ubuntu 20.04 LTS).

Cheers.

**Comment by doublemcz** [comment]
+1

**Comment by fungiboletus** [comment]
In addition to the flashing workarounds, I found a way to make a TCP tunnel of the serial-port from Windows to WSL2 which can be useful if you want to read/write on the serial port.

You can download [Null-modem emulator](https://sourceforge.net/projects/com0com/) and execute the `com2tcp-rfc2217.bat` in PowerShell on the right serial port and an available TCP port you like. That will make a TCP server connected to the serial port.

```powershell
.\com2tcp-rfc2217.bat COM3 5421
```

Then in Linux you can install `socat` and use it to connect to the TCP server and redirect the data to a virtual PTY. I didn't manage to connect to `localhost` from WSL2 to a Windows TCP Server but the IP on the local network works for me. You can find it using `ipconfig` and replace `192.168.1.10` in the following command.

```bash
socat -d -d tcp-connect:192.168.1.10:5421 pty,raw,echo=0
```

`socat` will print something like `PTY is /dev/pts/3` and then you can use `/dev/pts/3` as your virtual serial port.

**Comment by dsonyy** [comment]
@fungiboletus I have tested your workaround, and it seems like the communication is estabilished. However, I have a problem with baudrate setting. After running `com2tcp-rfc2217.bat`, it by default prints `... baud=19200 ...`. I would like change it but I haven't found any working solution yet.

**Comment by syamsidhardh** [comment]
I'm eagerly waiting for a proper fix for this issue. I hope MS will fix this soon.

**Comment by ocalvo** [comment]
Frustrated that I cannot do development of the [SMS integration for homeassistant](https://www.home-assistant.io/integrations/sms/) with WSL 2.

**Comment by daviddecobee** [comment]
Disappointed after scanning through here that I had to downgrade to WSL1, currently using Ubuntu 20.04.

**Comment by dandymon** [comment]
BUMP! For all the reasons other people have mentioned on this thread, with a specific echo of what @daviddecobee said!!

**Comment by adokitkat** [comment]
Please Microsoft make this a high priority...

**Comment by bnystrom** [comment]
Stuck in WSL 1 land also since I rely on WSL for doing microcontroller development and rely on USB communication with physical hardware.

**Comment by kingsumos** [comment]
This issue happens in other distros as well (e.g openSUSE 15.1/15.2).

**Comment by grizmin** [comment]
bump

**Comment by jjclose** [comment]
I'm in the same boat as most other people here.  After discovering this issue exists, I'm currently deciding whether to downgrade to WSL 1 or to switch to my Mac for development.

**Comment by rickiewars** [comment]
> I'm in the same boat as most other people here. After discovering this issue exists, I'm currently deciding whether to downgrade to WSL 1 or to switch to my Mac for development.

I've already switched to linux for all of my workflow due to this issue. I only rely on Windows for some games at the moment and honestly, I don't even miss my Windows installation. I know that for most users this won't be possible but for those who can, give it a chance!

**Comment by mkanet** [comment]
Is there any way to tell if there is any active development for implementing this capability?  Im currently using Oracle Virtualbox Linux VMs with USB filters for serial devices on Windows 10.

**Comment by kattaliraees** [comment]
This is a real bummer. Especially for hardware/firmware people. I am wondering if any work is happening for this issue. Looks like almost 2 years since this issue opened

**Comment by wgaylord** [comment]
So wait they can add GPU pass thru easily but can't add the far simpler USB pass thru? Referencing https://docs.microsoft.com/en-us/windows/wsl/wsl2-faq#can-i-access-the-gpu-in-wsl-2-are-there-plans-to-increase-hardware-support

**Comment by jjclose** [comment]
I commented on this previously so I thought I would follow up in order to provide feedback to Microsoft WSL team.  Since I badly need USB/COM ports in the Linux device setup, I was debating whether to revert back to WSL1 or use something else.  In the end I decided to move to an Ubuntu box and have been happy ever since, and frustrated that I spent longer than I should have waiting on WSL.   Sad to not have the parts of Windows that I was using, but those all have workarounds and the COM port issue does not.

**Comment by mkanet** [comment]
Is there a way to follow how far the WSL team has gotten in implementing USB/COM ports for WSL2 since July 2019?

**Comment by rickiewars** [comment]
> Is there a way to follow how far the WSL team has gotten in implementing USB/COM ports for WSL2 since July 2019?

I haven't seen anything mentioned about this issue in their roadmaps so I'm assuming that no one inside the Microsoft team has been working on this. Still I can't say anything for certain.

@jjclose, Always nice to see that more people are fully switching to Linux and rely less on Windows :)

**Comment by rajithavk** [comment]
I'm surprised why they are not implementing this or showing any interest at all, basically that's the most important feature I'd need as an embedded engineer to use WSL.

**Comment by xornet-sl** [comment]
Bump!!!

**Comment by qdx** [comment]
Bump!

**Comment by bojanpotocnik** [comment]
I agree that this feature is important, but there is no need to just "bump" this topic every 4 days and give hundreds of subscribed people fake hope with email notification. Considering that this is opened for almost 2 years, once per month is probably enough.

You can try writing an email to WSL Team as mentioned under [Additional information](https://github.com/microsoft/WSL/blob/master/CONTRIBUTING.md#user-content-7-additional-information) - maybe this will get enough attention to at least provide official response to this issue and plan on fixing it.

**Comment by jerrylogansquare** [comment]
please, pretty please!

**Comment by 24540477zxc** [comment]
WSL's IO performance is horribly low.If WSL2 does not support serial, it can only switch to linux completely

**Comment by slorquet** [comment]
I wonder if a tcp/serial bridge running on the host would help.

**Comment by florianhumblot** [comment]
@slorquet that does work, I'm currently using: https://github.com/abobija/idfx which is basically a TCP bridge between the host machine and WSL allowing for serial passthrough.
Though this example is specific for esp-idf and esp tools

**Comment by HStry** [comment]
@bojanpotocnik Seeing as this is a fairly basic request, and it's already almost 2 years open without any official response by Microsoft, I'd say it's more than valid for people to be extremely frustrated about this. Sure, there are workarounds for some use cases, but not for every case. And if you're working with embedded systems or COM port connected sensors, it's a definite show-stopper.

**Comment by jviskari** [comment]
Is there any plans to make this work?

**Comment by 24540477zxc** [comment]
Maybe Microsoft hasn’t figured it out yet

**Comment by tzuriely** [comment]
????????

**Comment by JosuGZ** [comment]
This is how I feel about it:

https://www.youtube.com/watch?v=BuLyxcyhbLQ

Damnit! Its sooo close!

BTW some talk about downgrading to wsl 1, is it possible? I've tried naively and now I get `Error response from daemon: open \\.\pipe\docker_engine_linux: The system cannot find the file specified.`

**Comment by github348** [comment]
Bump!!!!!!!!!!

**Comment by pmckeon** [comment]
This is a problematic issue for those of us doing embedded development on platforms like Arduino / Espressif as you can not program a device over USB serial port.

**Comment by MarkerViktor** [comment]
Bump!

**Comment by CrossMyHeart** [comment]
Actually it's very important and necessary for embedded software debug.
This issue is fatal to WSL2!!!!

**Comment by jjclose** [comment]
People, this has been going on for, literally, years.  I strongly recommend that you make a decision and move on.  I posted here and then quickly came to the realization that MS is not going to fix this, at least not on any reasonable timeline.  They are clearly aware of this issue, and are ignoring it, so it is pointless to keep posting about it.  I decided to ditch Windows and installed Linux, since the pain of what I am missing in Windows is not as bad as the pain of what I needed in Linux when working from Windows.  There are only two reasons that I can think of that explain Microsoft's silence on this:  either they really don't care about it and are abandoning the project, or they are not fixing this because they're working on something bigger regarding Linux.

**Comment by d-c-d** [comment]
Another alternative solution when a Linux system is not available is to use Cygwin, that's what I've done. Been using it from pre wsl days - just another tool

**Comment by Trance-Paradox** [comment]
Really I don't know why this fix is taking so long. Also am not aware of any conspiracy theories but M$ can bring GPU accelerated GUI and application support in WSLg but can just share a COM port in wsl2??? Its now even a new feature it was working in WSL1 is what I am aware off but even after so much time has passed no response from M$ devs. I don't know if M$ developer think this feature is not worthy or just want embedded system developers either to buzz of to native Linux or use the oversized MS built tools???

**Comment by nbdd0121** [comment]
> Really I don't know why this fix is taking so long.

This is far from trivial to implement. They'll likely need to implement a kernel module to support this. You can't even easily access a serial port on another Linux device as if it's a local serial port, needless to say about a serial port on Windows host.

Support it in WSL1 is much easier because it's only userspace APIs to replicate, no kernel module is needed.

Sending angry comments or meaningless bumps will not help the scenario and will only annoy readers. Many people subscribe to the issue for updates not complaints.

**Comment by jjclose** [comment]
> > Really I don't know why this fix is taking so long.
> 
> This is far from trivial to implement. They'll likely need to implement a kernel module to support this. You can't even easily access a serial port on another Linux device as if it's a local serial port, needless to say about a serial port on Windows host.
> 
> Support it in WSL1 is much easier because it's only userspace APIs to replicate, no kernel module is needed.
> 
> Sending angry comments or meaningless bumps will not help the scenario and will only annoy readers….

This is a silly response.  People have been making this request for 2-3 years and getting absolutely zero response from Microsoft.  It is very necessary and core functionality.  Asking for a basic feature that still hasn’t been delivered annoys other readers?  Probably not as annoyed as we are at lacking this feature.  In addition, WSL2 already embeds an actual Linux kernel as opposed to a translation layer.  There is obviously a plethora of other serial port driver code available for Linux, so it’s hard to imagine that this would be such a daunting task for MS as you describe.

**Comment by rubberduck203** [comment]
The difficulty is likely due to WSL2 sitting on top of virtualization, unlike WSL1.

**Comment by nbdd0121** [comment]
> This is a silly response. People have been making this request for 2-3 years and getting absolutely zero response from Microsoft. It is very necessary and core functionality. Asking for a basic feature that still hasn’t been delivered annoys other readers? Probably not as annoyed as we are at lacking this feature. In addition, WSL2 already embeds an actual Linux kernel as opposed to a translation layer. There is obviously a plethora of other serial port driver code available for Linux, so it’s hard to imagine that this would be such a daunting task for MS as you describe.

I said *angry comments or meaningless bumps* annoy readers. Question marks and bumps serve absolutely no purpose in helping the scenario. If I work for WSL2 I'll completely disregard these comments.

Don't get me wrong, I absolutely want this feature, as many people do. I want it so bad that I even tried to implement this feature myself, and only then I learned that this task is more complicated than I expected. Kernel abstractions provided are very low-level for real devices and not for a virtual one like what is needed for this feature. The closest I can get is to go via pty but you cannot set baud rate via pty. And they'll also need to dynamically create and destroy these devices when COM ports are connected and disconnected from the host.

It is certainly doable given enough manpower, but it is far from trivial. Without infinite resource to implement all feature requests, I can understand why they prioritize features like WSLg which is wanted by more users.

**Comment by HStry** [comment]
> Don't get me wrong, I absolutely want this feature, as many people do. I want it so bad that I even tried to implement this feature myself, and only then I learned that this task is more complicated than I expected. Kernel abstractions provided are very low-level for real devices and not for a virtual one like what is needed for this feature. The closest I can get is to go via pty but you cannot set baud rate via pty. And they'll also need to dynamically create and destroy these devices when COM ports are connected and disconnected from the host.

I've not had the time to look into this much yet, but I've been thinking of trying to make a com-over-ip kind of thing, where a driver on Windows would just transparently forward all information to and from a com port to the local network of WSL, where that has a driver that does the inverse. I've only looked for a few minutes into it, but I think I saw a serial-over-ip thing somewhere. But I'm currently in a bit of a home reconstruction, so...

**Comment by siebert** [comment]
I found this workaround, which seems to be promising (but I have not tried it yet): https://matevarga.github.io/esp32/m5stack/esp-idf/wsl2/2020/05/31/flashing-esp32-under-wsl2.html

**Comment by void-spark** [comment]
Wonder if @craigloewen-msft could let us know if this is even remotely on their radar :)

**Comment by HStry** [comment]
> It this only a problem for 18.04 like the heading says. I think this is a problem for all the distros

Good point, didn't even see that. I'm using Debian, and same issue there.
@strike5150 could you make the heading perhaps less distro-specific? Maybe that'll make the scope of the issue clearer.

**Comment by craigloewen-msft** [comment]
Serial support in WSL 2 is definitely on our radar! Our current and immediate roadmap is focusing on finalizing tools like `--install`, `--mount`, GUI app support, and GPU support, but we will be taking a look at prioritizing this issue again as we go ahead with scheduling in the future. I'll be sure to post any updates to this Github thread, thank you all for the feedback and interest.

**Comment by GregWoods** [comment]
I for one appreciate the response from someone from Microsoft... albeit after 2 years!

It is frustrating though, when useless but headline grabbing features like "GUI App Support", and "GPU Support" take priority over getting the foundation of the product finished.

I can only conclude the masterplan is to move Windows onto a Linux kernel. Well, maybe my grandkids will benefit!

**Comment by CalvinXu17** [comment]
Today is 2021.9.16. This problem remains unsolved even after 2 years !!!

**Comment by IgnusG** [comment]
> I for one appreciate the response from someone from Microsoft... albeit after 2 years!
> 
> It is frustrating though, when useless but headline grabbing features like "GUI App Support", and "GPU Support" take priority over getting the foundation of the product finished.
> 
> I can only conclude the masterplan is to move Windows onto a Linux kernel. Well, maybe my grandkids will benefit!

GPU support is very much appreciated since it helps running ML algorithms on WSL. Just because it doesn’t fit the use case of this thread doesn’t mean it should be discarded in the trash like a mere buzzword gig. 

And yeah, I would also love to have support for working with embedded devices, but I love the work the MS team put into WSLg just as well.

**Comment by GregWoods** [comment]
@IgnusG  Thanks for correcting me regarding the importance of GPU support. My post was born out of frustration, and perhaps I was taking my general Microsoft grievances out on the WSL team!

**Comment by HStry** [comment]
@IgnusG, @ShaneEverittM, @GregWoods,
Yes, GPU support has real uses, however I would imagine PCIe passthrough or high performance gpu emulation to be magnitudes more complex than serial passthrough. Maybe something like a [serial-over-ethernet](https://en.wikipedia.org/wiki/Serial_over_LAN) kind of implementation where the communication of a serial port is transparently routed through the NAT system already in place for WSL. It's a low speed, well known, protocol. In fact, IPMI has something like that already: [_"v2.0, published on February 12, 2004: added features including Serial over LAN, Group Managed Systems, Enhanced Authentication, Firmware Firewall, and VLAN Support"_](https://en.wikipedia.org/wiki/Intelligent_Platform_Management_Interface#Version_history)
A quick search here on Github also shows several serial-over-ethernet projects.

@craigloewen-msft Seriously, it's rather embarrassing that Microsoft can't seem to figure this out, whereof large sections have already been worked out by others. Yes, GUI support is important, I'm not trying to undermine it, but it's been **_years!!_** You can't seriously expect us to believe this issue is _that_ hard. Heck, maybe talk to supermicro who know how to implement IPMI containing SoE since 2004, or a company actually named "serial-over-ethernet", maybe they could figure this out?

[host serial] <--> [WSL host IF] <~nat~> [WSL guest IF] <--> [guest serial]

Get it done. This should not take years to implement.

**Comment by wgaylord** [comment]
Worst part is it worked in WSL1 so they know HOW to do it but just aren't.

**Comment by jmccusker** [comment]
Given all of the uproar, I suspect the reason why WSL2 doesn't support USB/Serial ports has to do with some kind of security flaw.  Just my guess, but MS's silence on the issue seems to point in that direct.

**Comment by SunJun8** [comment]
> Hi @craigloewen-msft, it looks like you released a WSL preview for Windows 11 with serial USB support. Can you confirm that you include this feature ? Is this preview available for Windows 10 ? If not, when will it be available ? Thanks for your feedback !

The subsystem in the Microsoft Store requires that the system version must be above 22000, which means it must be win11...maybe

**Comment by jmccusker** [comment]
Requiring Windows 11 for WSLg is getting annoying. There's no reason for
this other than marketing pressure.

**Comment by benhillis** [comment]
@jmccusker - actually, you're wrong. The shared memory infrastructure that WSLg uses is currently only available on Windows 11.

**Comment by wgaylord** [comment]
So wait, GPU pass thru doesn't require shared memory but Serial devices do? Sounds a bit odd / untrue.

**Comment by zt-luo** [comment]
i have noticed than the latest Linux kernel version 5.10.60.1 have support USB over IP，anyone know how to use it？

Release Note：

> Update Linux kernel to 5.10.60.1
Enable additional USB kernel configuration options for interacting with an Arduino over USB
Update to upstream stable kernel release 5.10.60
Enable virtio-pmem with support for PCI BAR-relative addresses
Enable vPCI support under Hyper-V for ARM64
Enable io_uring support
Enable USB over IP support
Enable paravirtualized spinlock support for x86_64
Refresh dxgkrnl driver to pick up bug fixes and code cleanups
Enable NFS client support for NFSv4.1* Improve the error message when 9p or fstab mounts fail

**Comment by kattaliraees** [comment]
From the release notes (store version of WSL)

"- Enable additional USB kernel configuration options for interacting with an Arduino over USB"

Have any body tried this on a windows 11 with WSL installed from the store

**Comment by lsthiros** [comment]
[Looks like a Microsoft intern came to the rescue!](https://devblogs.microsoft.com/commandline/connecting-usb-devices-to-wsl/) Thanks @Nelson Daniel Troncoso Aldas!

The link above is to the Microsoft dev blog where they've outlined how to get around this frankly embarrassing, three year old, critical, regression. "The device stack is no longer interoperable" is a showstopper in any low level emulation software, and it's baffling that WSL2 was shipped with such a glaring issue.

**Comment by JosuGZ** [comment]
> [Looks like a Microsoft intern came to the rescue!](https://devblogs.microsoft.com/commandline/connecting-usb-devices-to-wsl/) Thanks @NelsonDaniel !
> 
> The link above is to the Microsoft dev blog where they've outlined how to get around this frankly embarrassing, three year old, critical, regression. "The device stack is no longer interoperable" is a showstopper in any low level emulation software, and it's baffling that WSL2 was shipped with such a glaring issue.

"first ensure you’re running Windows 11" :(

**Comment by felixstorm** [comment]
> "first ensure you’re running Windows 11" :(

I have not tried it on Windows 10 (yet), but the corresponding wiki article in usbipd-win only mentions the kernel version 5.10.60.1, but not word of Windows 11: https://github.com/dorssel/usbipd-win/wiki/WSL-support  
And it even explains how to compile your own kernel with USB/IP support in case you‘re on a lower version.

**Comment by AFellmett** [comment]
> "first ensure you’re running Windows 11" :(

I can confirm, it's running on Windows 10 as well. The only "drawback" seems to be the need for a custom kernel, cause common usb serial drivers (like cp210x and ch34x) are not available out of the box.
Nevertheless, the process as described in the wiki as mentioned by @felixstorm is worth reading and helps alot bringing the long requested feature to wsl.

**Comment by lsthiros** [comment]
Bah, I bit the bullet and did the windows 11 update only to find that the included WSL2 kernel doesn't have the modules for the SiLabs bridges littering desk. Can confirm though, that my WSL2 instance notices and acknowledges the enumeration of the UART bridge.

That being said, this is a workaround, and the issue shouldn't be closed until WSL2 once again supports the USB pass-through out of the box.

**Comment by kumekay** [comment]
> Bah, I bit the bullet and did the windows 11 update only to find that the included WSL2 kernel doesn't have the modules for the SiLabs bridges littering desk. Can confirm though, that my WSL2 instance notices and acknowledges the enumeration of the UART bridge.
> 
> That being said, this is a workaround, and the issue shouldn't be closed until WSL2 once again supports the USB pass-through out of the box.

It's not hard to compile a custom kernel with support for cp210*. Here are mine instructions https://kumekay.com/compiling-custom-kernel-for-wsl2/

**Comment by lkoepsel** [comment]
I want to update my earlier post...
1) When I attempted to use WSL2 with a serial port to program the UNO, I did NOT follow the advice left by the nice Microsoft intern.  Not a criticism, simply a note for others. BTW - my NUC is not a Windows 11 compliant computer...I am using W11 as part of the early access program.

2) I have also noticed that my access to the serial port from WSL2 requires initial access from Windows. For example, if I haven't used the serial port since I last rebooted, I am unable to access the port in WSL2 via /dev/ttyS3 (for COM3). Once I "touch" the port by uploading a file via the Arduino IDE or even open the port using PuTTY (no interaction, simply open a window for Serial port COM3), I'm able to use the port in WSL2. In my case in WSL2, I'm using averdude -P /dev/ttyS3 ...


Hope this helps someone else.

**Comment by onomatopellan** [comment]
@lkoepsel Are you sure you are running a WSL2 distro? What's the output for `wsl.exe -l -v` ?

**Comment by lkoepsel** [comment]
Well...good question! I ran the above command in my WSL? box and got the following:
  NAME      STATE           VERSION
* Ubuntu    Running         1

I'm assuming the Version of 1 says that I'm running WSL1?

If so, does WSL1 run in Windows 11?
Is it possible to have both versions exist in Windows 11? As I'm "pretty sure" I also installed WSL2 via the W11 store.

Of course, I'm no longer sure of anything. :)

**Comment by onomatopellan** [comment]
@lkoepsel Correct, both WSL1 and WSL2 distros can run in Windows 11. You default is probably WSL1 so anything you install from the Store will become a WSL1 distro. You need to run `wsl.exe --set-default-version 2` at least once in order to get WSL2 distros.

**Comment by lkoepsel** [comment]
@onomatopellan, thanks for the help! 

I'm testing this for a Community College curriculum, which means I need to make it somewhat fail safe...and this might help someone else.

I did the following and there wasn't a difference between WSL1 and WSL2 in the requirement to "open" the port via a Windows app prior to using it in WSL2.

* I rebooted
* opened WSL
* checked status and it is WSL1
* set the default WSL to 2 and checked status, its WSL2
* attempted to upload to Uno, failed
* ran the Arduino IDE in Windows and uploaded a file to the UNO
* went back to WSL2 and  uploaded, success

**Is there supposed to be a difference between the two WSL version as it pertains to the serial port?**

Actual terminal session:
```
lkoepsel@DESKTOP-KPLPT00:~$ wsl.exe --status
Default Distribution: Ubuntu
Default Version: 1
Please enable the Virtual Machine Platform Windows feature and ensure virtualization is enabled in the BIOS.
For information please visit https://aka.ms/enablevirtualization
WSL version: 0.50.2.0
Kernel version: 5.10.74.3
WSLg version: 1.0.29
Windows version: 10.0.22000.346
lkoepsel@DESKTOP-KPLPT00:~$ wsl.exe --set-default-version 2
For information on key differences with WSL 2 please visit https://aka.ms/wsl2
The operation completed successfully.
lkoepsel@DESKTOP-KPLPT00:~$ wsl.exe --status
Default Distribution: Ubuntu
Default Version: 2
Please enable the Virtual Machine Platform Windows feature and ensure virtualization is enabled in the BIOS.
For information please visit https://aka.ms/enablevirtualization
WSL version: 0.50.2.0
Kernel version: 5.10.74.3
WSLg version: 1.0.29
Windows version: 10.0.22000.346
lkoepsel@DESKTOP-KPLPT00:~$ cd 10c_labs_class/Labs/1B/1_blink/
lkoepsel@DESKTOP-KPLPT00:~/10c_labs_class/Labs/1B/1_blink$ make flash
avrdude -c Arduino -p atmega328p -F -V -P /dev/ttyS3 -b 115200   -U flash:w:main.hex
avrdude: stk500_recv(): programmer is not responding
avrdude: stk500_getsync() attempt 1 of 10: not in sync: resp=0x00
^Cmake: *** [Makefile:140: flash] Interrupt

#################################
# AT THIS POINT I RAN THE ARDUINO IDE
#################################

lkoepsel@DESKTOP-KPLPT00:~/10c_labs_class/Labs/1B/1_blink$ make flash
avrdude -c Arduino -p atmega328p -F -V -P /dev/ttyS3 -b 115200   -U flash:w:main.hex

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.00s

avrdude: Device signature = 0x1e950f (probably m328p)
avrdude: NOTE: "flash" memory has been specified, an erase cycle will be performed
         To disable this feature, specify the -D option.
avrdude: erasing chip
avrdude: reading input file "main.hex"
avrdude: input file main.hex auto detected as Intel Hex
avrdude: writing flash (684 bytes):

Writing | ################################################## | 100% 0.12s

avrdude: 684 bytes of flash written

avrdude: safemode: Fuses OK (E:00, H:00, L:00)

avrdude done.  Thank you.

```

**Comment by onomatopellan** [comment]
@lkoepsel You need to convert the Ubuntu WSL1 to WSL2 or you won't notice any difference.
`wsl.exe --set-version Ubuntu 2`

**Comment by lkoepsel** [comment]
@onomatopellan Thank you for the explanations and the time.

I used
```
 wsl.exe --set-version Ubuntu 2
```
and the WSL crashed each time.

I have some other work to do which involves rolling back to Windows 10, so for now, I'll leave this issue alone. I (and I'm sure others) appreciate the clarification.

**Comment by onomatopellan** [comment]
@lkoepsel those crashes are probably because of the "Please enable the Virtual Machine Platform Windows feature and ensure virtualization is enabled in the BIOS. " message you got when you did enable WSL2.

**Comment by Sourabh1107** [comment]
Hi All,

I am still using Windows 10 (as currently my company IT team is not ready to upgrade to W11) with build version 1909 and OS build 18368.1916 I have docker for Windows with Linux containers and WSL2 enabled. My issue is I am not able to mount and connect USB devices to access from docker container. Since recommended way is to upgrade to W11 I am looking for some workaround. I basically want to format, partition and flash SD card connected through microSD USB reader.

I found the [Dynamically access USB devices](https://stackoverflow.com/a/62758958/15610191) which described accessing USB devices but i tried and failed with this. I think it doesn't mount device first. I am stuck and don't know how to proceed further.

Please someone let me know how we can do this for older version of Windows.

**Comment by devcui** [comment]
![image](https://user-images.githubusercontent.com/26159886/148773730-059662f2-dd25-4ab0-a1cb-6b8523196716.png)

the same problem

**Comment by lkoepsel** [comment]
As @devcui points out, the version of WSL that is running is Version 2. I have not been successful running Version 2 with serial access. I have been successful running Version 1 with serial access. This has been on Windows 10.

I recommend making the same attempts above, however using WSL Version 1.

**Comment by devcui** [comment]
@lkoepsel  thanks for your answer, i find a document that in https://docs.microsoft.com/en-us/windows/wsl/connect-usb, butdisappointingly, it doesn't work successful. and now i will try WSL Version 1.

**Comment by nullromo** [comment]
> https://docs.microsoft.com/en-us/windows/wsl/connect-usb it doesn't work successful.

This article worked for me on Windows 11 with WSL 2. 🎉

**Comment by elasticdotventures** [comment]
This recent blog post from microsoft explains:
https://devblogs.microsoft.com/commandline/connecting-usb-devices-to-wsl/

**Comment by J-Rios** [comment]
```text
This functionality takes advantage of the [USB/IP protocol](https://www.kernel.org/doc/html/latest/usb/usbip_protocol.html) to forward USB traffic over the virtual network interface between Windows and WSL.
```

It makes me sad, that current way to handle this is using that "hacky" workaround, and I'm affraid that it could makes some troubles with windows firewall at some point...

Is there any planning on WSL roadmap regarding this for a more directly compatible solution that doesn't require this workaround?

**Comment by jrichrd** [comment]
> > This functionality takes advantage of the [USB/IP protocol](https://www.kernel.org/doc/html/latest/usb/usbip_protocol.html) to forward USB traffic over the virtual network interface between Windows and WSL.
> 
> It makes me sad, that current way to handle this is using that "hacky" workaround, and I'm affraid that it could makes some troubles with windows firewall at some point...
> 
> Is there any planning on WSL roadmap regarding this for a more directly compatible solution that doesn't require this workaround?

+1 to J-Rios' comment. For embedded developers who want to develop on the Windows platform vs native *nix/OSX this is not an acceptable solution. Is someone from Microsoft able to provide details on why this is a difficult problem to solve and integrate into WSL2? I'm currently using WSL1 to get around this issue but that means dealing with other issues e.g. extremely slow calls into files on the mounted disk and network calls e.g. pip, npm calls take 10x the time from a WSL2 installation. I appreciate the complexity in solving this from an engineering perspective but would be great to see a roadmap to help folks plan e.g. I don't want to use a Mac, I enjoy Windows, but working around these issues isn't a great use of time :)

**Comment by dobromyslov** [comment]
USBIPD works good in Windows 10 with WSL 2 kernel  5.10.60.1 as described here https://docs.microsoft.com/en-us/windows/wsl/connect-usb

The kernel sees my serial adapter HL-340 but does not bind it to the `/dev/ttySx` file.
Looks like the kernel driver for HL-340 is missing as @AFellmett  described in https://github.com/microsoft/WSL/issues/4322#issuecomment-962599648

Will try to rebuild the kernel with HL-340 support as described here: https://github.com/dorssel/usbipd-win/wiki/WSL-support#building-your-own-usbip-enabled-wsl-2-kernel

Can WSL maintainers update the prebuilt kernel with more serial adapters support?

```bash
# lsusb
Bus 001 Device 002: ID 1a86:7523 QinHeng Electronics HL-340 USB-Serial adapter.

# dmesg
[ 9391.193570] vhci_hcd vhci_hcd.0: pdev(0) rhport(0) sockfd(3)
[ 9391.193572] vhci_hcd vhci_hcd.0: devid(65537) speed(2) speed_str(full-speed)
[ 9391.193658] vhci_hcd vhci_hcd.0: Device attached
[ 9391.469817] vhci_hcd: vhci_device speed not set
[ 9391.539507] usb 1-1: new full-speed USB device number 2 using vhci_hcd
[ 9391.629431] vhci_hcd: vhci_device speed not set
[ 9391.699427] usb 1-1: SetAddress Request (2) to port 0
[ 9391.733154] usb 1-1: New USB device found, idVendor=1a86, idProduct=7523, bcdDevice= 2.64
[ 9391.733156] usb 1-1: New USB device strings: Mfr=0, Product=2, SerialNumber=0
[ 9391.733158] usb 1-1: Product: USB Serial
```

**Comment by dobromyslov** [comment]
I confirm that kernel rebuild helps and commonly used Chinese USB-to-TTL serial adapters work if the following configuration is added to the kernel at `Device Drivers -> USB Support -> USB Serial Converter Support`:
* USB Winchiphead CH341 (CONFIG_USB_SERIAL_CH341)
* USB CP210x family (CONFIG_USB_SERIAL_CP210X)
* USB Prolific 2303 (CONFIG_USB_SERIAL_PL2303)

These drivers MUST be built-in (marked as [ * ]). Nothing works if I build them as modules (mark as [ M ]). Looks like WSL does not understand that it should load a kernel module when a usb device is attached.

Another caveat was found. During flashing procedure the USB device may be hard resetted by the flashing tool and disconnect-connect occurs. WSL does not reattach the USB device, the serial connection breaks and the flasher just hangs. Looks like we need a feature to automate the USB device reattach on the disconnect-connect event.

**Comment by benjjyman** [comment]
+1,
Having to switch back to windows everytime I want to flash or dump is a pain in the A.
please fix

**Comment by AlejandroFNadal** [comment]
I just come to say that I am also in need of this feature, I hope that microsoft solves this

**Comment by benjjyman** [comment]
> I just come to say that I am also in need of this feature, I hope that microsoft solves this

Me too!

**Comment by mrtreasurer** [comment]
Also interested in this feature

**Comment by sergiunagy** [comment]
Would be a nie feature to be supported natively. 
Doing both embedded programming and working with Docker I would like to have WSL2 natively support both.

**Comment by kattaliraees** [comment]
😑😑😑

https://developer.apple.com/documentation/virtualization/running_gui_linux_in_a_virtual_machine_on_a_mac

https://developer.apple.com/documentation/virtualization/serial_ports

https://developer.apple.com/videos/play/wwdc2022/10002/

**Comment by ghmanoj** [comment]
> I confirm that kernel rebuild helps and commonly used Chinese USB-to-TTL serial adapters work if the following configuration is added to the kernel at `Device Drivers -> USB Support -> USB Serial Converter Support`:
> 
> * USB Winchiphead CH341 (CONFIG_USB_SERIAL_CH341)
> * USB CP210x family (CONFIG_USB_SERIAL_CP210X)
> * USB Prolific 2303 (CONFIG_USB_SERIAL_PL2303)
> 
> These drivers MUST be built-in (marked as [ * ]). Nothing works if I build them as modules (mark as [ M ]). Looks like WSL does not understand that it should load a kernel module when a usb device is attached.
> 
> Another caveat was found. During flashing procedure the USB device may be hard resetted by the flashing tool and disconnect-connect occurs. WSL does not reattach the USB device, the serial connection breaks and the flasher just hangs. Looks like we need a feature to automate the USB device reattach on the disconnect-connect event.

I will try this…

**Comment by nicoande** [comment]
Thread opened 3 Years ago and still nothing ?
Guess I have to switch to macOS then...
Please make it happen ! It's a must !

**Comment by jjclose** [comment]
Or VirtualBox and a Linux distro, as I did

**Comment by nullromo** [comment]
@benjjyman @AlejandroFNadal @mrtreasurer @sergiunagy @ghmanoj @nicoande @jjclose please refer to [this comment](https://github.com/microsoft/WSL/issues/4322#issuecomment-1027365578). Hopefully that will help solve your current issues. Furthermore, before posting a "me too" comment on a thread with 101 participants, think about how many e-mails you will generate.

**Comment by jjclose** [comment]
> 

@nullromo as many people have already pointed out, that solution did not work for them;  it is also not really a satisfactory approach since it is sort of a hack.  rather than scold people here for replying (it's bad to reply on a thread about a shared problem???). maybe you could support the community concern that this issue has not been addressed in three years.

**Comment by wenbozzz** [comment]
I don't think it's possible to **share** a USB device across Windows and WSL2 (like you would be able to do in WSL1), either the host owns the USB device or it's passed through to WSL2 (like other VMs running fully fledged Linux kernel).
But maybe providing a GUI for user to select which USB device they want to use with WSL2 is not a bad idea?🤔

**Comment by rubberduck203** [comment]
> Guess I have to switch to macOS then...

The situation is just as bad on Mac. 
The virtual box solution is iffy at best, as it tends to require seemingly magic combinations of VB & Docker versions, as well as only working for certain serial devices.

After years of looking for a satisfactory solution I can safely say that the only one that works consistently is running a Linux distro on your metal.

**Comment by rickiewars** [comment]
> I don't think it's possible to **share** a USB device across Windows and WSL2 (like you would be able to do in WSL1), either the host owns the USB device or it's passed through to WSL2 (like other VMs running fully fledged Linux kernel). But maybe providing a GUI for user to select which USB device they want to use with WSL2 is not a bad idea?🤔

This would be the solution I think and has been available in VirtualBox and KVM for years. Even through a command I should be happy with because it will be scriptable then. 
Because the windows part is close source, Microsoft has to implement i themselves but they always can take some inspiration from KVM because the linux kernel is open source. It won't be an 1-to-1 comparison but I can't imagine the example is of any help at all.

I've switched to Linux full-time and virtualize Windows in kvm in case I need it because it seems that that gives me more options.

**Comment by milomai** [comment]
just an idea:
in windows: turn serial to tcp (like com0com)
in linux: turn tcp to serial (like ser2net)

**Comment by charterchap** [comment]
The answer is in the post by @dobromyslov 
to abbreviate:
```
ADMIN PS> wsl --update

PS> winget install --interactive --exact dorssel.usbipd-win

WSL> sudo apt install linux-tools-5.4.0-77-generic hwdata
WSL> sudo update-alternatives --install /usr/local/bin/usbip usbip /usr/lib/linux-tools/5.4.0-77-generic/usbip 20

# Open a new ADMIN PS
PS> usbipd wsl list
PS> usbipd wsl attach --busid <busid>

WSL> lsusb 
WSL> ll /dev/ttyUSB*
PS> usbipd wsl detach --busid <busid> # when you're done with your device
```

**Comment by ackshooerry** [comment]
Despite the finality of the most recent response, I don't think this issue is settled at all. I have followed all of the steps outlined here. I can see and attach my USB to serial adapter in PowerShell:

```
PS C:\WINDOWS\system32> usbipd wsl list
BUSID  VID:PID    DEVICE                                                        STATE
1-3    2c7c:0125  HighSpeed USB Composite Device (0008), Quectel USB DM Por...  Not attached
1-4    0bda:0129  Realtek USB 2.0 Card Reader                                   Not attached
1-5    067b:2303  Prolific USB-to-Serial Comm Port (COM8)                       Attached - Ubuntu-20.04
1-7    0bda:d723  Realtek Bluetooth Adapter, Realtek 8723DU Wireless LAN 80...  Not attached
1-8    058f:5608  USB 2.0 Web Camera                                            Not attached
```

I can see the USB device in WSL:
```
prompt:/dev$ lsusb
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 002: ID 067b:2303 Prolific Technology, Inc. PL2303 Serial Port
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
```
  
But when I list my TTY devices, I get:
```
prompt:/dev$ ls tty*
tty    tty11  tty15  tty19  tty22  tty26  tty3   tty33  tty37  tty40  tty44 tty48  tty51  tty55  tty59  tty62  tty9   ttyS3
tty0   tty12  tty16  tty2   tty23  tty27  tty30  tty34  tty38  tty41  tty45  tty49  tty52  tty56  tty6   tty63  ttyS0
tty1   tty13  tty17  tty20  tty24  tty28  tty31  tty35  tty39  tty42  tty46  tty5   tty53  tty57  tty60  tty7   ttyS1
tty10  tty14  tty18  tty21  tty25  tty29  tty32  tty36  tty4   tty43  tty47  tty50  tty54  tty58  tty61  tty8   ttyS2
prompt:/dev$ 
``` 

However, as seen above, this device is assigned to COM8 in Windows, so I would expect to see it as ttyS8 in WSL. Unfortunately, it isn't listed. As I read it, this is literally the exact same issue that was initially posted about here 3+ years ago.

I am running WSL2 on Windows 10, Ubuntu 20.04, Linux Core is 5.10.106.1.

**Comment by charterchap** [comment]
It will appear as `ls -la /dev/ttyUSB*` and will not correlate to the windows device in any way. The first you attach will probably show as `/dev/ttyUSB0`  although you clearly aren't getting that either. You can also try `dmesg|grep tty` to see if it has a device

**Comment by Eplankton** [comment]
> 

I have exact same problem, have you found any solution yet? Please help me.

**Comment by jrj2211** [comment]
@ackshooerry I also use prolific serial adapters, but from what I've tried the driver isn't included by default in the kernel. I did rebuild the kernel once and got my prolific adapters to work, but have since just bought FTDI USB serial cables as they work out of the box. Not sure if you did rebuild the kernel, or if that is these drivers are now included by default (last I tried was a few months ago).

**Comment by adeturner** [comment]
+1 Same issue - following the Pi pico install.
COM5 is not mapped to /dev/ttyS5, so minicom cant connect to it in WSL
I tried usbipd, lsusb sees the device - no tty though

```bash
$ lsusb
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 002: ID 2e8a:0003 Raspberry Pi RP2 Boot
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub

$ ls -l /dev/ttyU*
ls: cannot access '/dev/ttyU*': No such file or directory

$ ls -l /dev/ttyS*
crw------- 1 root root 4, 64 Jan 14 10:08 /dev/ttyS0
crw------- 1 root root 4, 65 Jan 14 10:08 /dev/ttyS1
crw------- 1 root root 4, 66 Jan 14 10:08 /dev/ttyS2
crw------- 1 root root 4, 67 Jan 14 10:08 /dev/ttyS3
```

**Comment by ghmanoj** [comment]
Hey everyone… please stop replying in this thread its sending messages to everyone that is mentioned in the thread. 
we do hope this issue gets resolved but due to fundamental limitations of WSL it won’t be fixed. Let’s just use Linux rather than an emulator?

**Comment by pwfff** [comment]
+1 here as well. Can't get my Prolific device to show up as `/dev/ttyUSB*` even though it's in `lsusb`. Guess I'm buying an FTDI cable because I've already spent too long on this and I sure ain't about to recompile my kernel for this...

**Comment by lsthiros** [comment]
This has been open for four years, rattling my inbox nonstop. Microsoft devs either needs to confirm that this is an intended major-feature removal and close the issue, or classify it as a major regression with known workaround (USBIP).

The silence on such a key feature is unacceptable, and the level disrespect towards the developer community shown through four years of silence is unbelievable.

Can we _please_ have a triage statement from devs on where this issue stands, or at least what would be required from the community to develop a POR fix?

**Comment by lemonking77** [comment]
We really need this feature

**Comment by mkanet** [comment]
**Just a note:**   Not everyone here has the skillset and/or motivation to jump through hoops installing usbipd-win, build their own USB/IP-enabled WSL 2 kernel, and try to get them to talk to each other.  

I'm not even a developer... I'm just a regular guy who would like their USB dongle to actually work after installing [Home Assistant](https://www.home-assistant.io/).  Currently, using it as VMWare VM on Win11.

**Comment by bwims** [comment]
My guess is that Microsoft is using this as a lever to get people to upgrade to Win11

https://learn.microsoft.com/en-us/windows/wsl/connect-usb

`Windows 11 is recommended for connecting USB devices to a Linux distribution running on WSL 2. However, Windows 10 can be used to connect USB devices by building your own USBIP enabled WSL 2 kernel following the [instructions in the USBIPD-WIN project repo](https://github.com/dorssel/usbipd-win/wiki/WSL-support).`

**Comment by supleed2** [comment]
Even if it is, in both cases once you have a compatible kernel, you still need to use the `usbipd-win` project, which means running a command to pass through the device every time. It's not automatic at all as it was with WSL1. This is an issue on Win11 too.

**Comment by bwims** [comment]
> Even if it is, in both cases once you have a compatible kernel, you still need to use the `usbipd-win` project, which means running a command to pass through the device every time. It's not automatic at all as it was with WSL1. This is an issue on Win11 too.

I agree, but the point is, I'd be surprised if full USB support ever came to Win10, even if it eventually arrives for Win11.  Yes, I am cynical but I do have reasons...

**Comment by slorquet** [comment]
Of course it is.

Who actually believes Microsoft is interested in Linux for the sake of advancing the free software community?

**Comment by koffes** [comment]
I have read the entire thread, and I also join the bandwagon and ask if we can please get any timeline or news regarding this feature. The USBIPD is a nice workaround, but no replacement for native USB forwarding and multi-device auto-attaching which is present in VMware and VirtualBox.

**Comment by bwims** [comment]
I've now installed VMWare, and it works perfectly well for my needs (Docker and access to USB devices)
EVEN BETTER, it uses NAT and DHCP to get a SENSIBLE IP address!

Sadly, it appears WSL1 and WSL2 are a waste of time for different reasons.

**Comment by jjclose** [comment]
I contributed to this thread with one of the earliest complaints.  I think
it was nearly two years ago.  It's sad to see that Microsoft cares so
little about this issue and it's confusing to me why they would
de-prioritize it.  Back then, I came to the same conclusion as you and just
installed linux on VirtualBox.

**Comment by adalava** [comment]
I was enjoying WSL but I gave up and moved to Linux completely. I still have Windows installed on another partition since some tools are Windows only, but most Windows applications that use serial port are so simple that run fine under Wine on Linux.

**Comment by shadrqen** [comment]
https://learn.microsoft.com/en-us/windows/wsl/connect-usb

**Comment by jviskari** [comment]
Still workaround.

**Comment by compnski** [comment]
I also gave up on windows and just use linux full time. https://www.codeweavers.com/crossover has been helpful in running any last software that regular wine didn't work with.

**Comment by crose72** [comment]
I actually followed the steps here https://learn.microsoft.com/en-us/windows/wsl/connect-usb and I had it working for both windows 10 and 11 from what I can remember.  I haven't needed to connect my FTDI serial to USB converter to WSL2 for a while and I just followed my notes and it's no longer working.  I can attach the device to WSL2, but it no longer shows up in /dev/ttyUSB0 anymore.  How do I fix that?

**Comment by Clockwork-Muse** [comment]
@crose72 - You'll want/need to check the [usbipd repo](), likely.

Note that there are a number of issues with some specific serial chips, [like this one](https://github.com/dorssel/usbipd-win/issues/59), many of which stem from the default kernel not containing certain drivers - you may need to compile one yourself.

**Comment by Clockwork-Muse** [comment]
I'm the wrong person to ask, unfortunately.

**Comment by crose72** [comment]
@Clockwork-Muse no worries, I'll try your suggestion all the same.  Are there any FTDI to USB devices that you know work?  If so I wouldn't mind getting one of those as a quick solution.

**Comment by mrv96** [comment]
It seems that we have a solution for the issue of @crose72 (and mine): https://github.com/dorssel/usbipd-win/issues/948#issuecomment-2138336528

**Comment by OsamaAtef2025** [comment]
> * رقم إصدار Windows الخاص بك: Microsoft Windows [الإصدار 10.0.18932.1000]
> * ما تفعله وما يحدث:
> 
> 1. تشغيل WSL2 Ubuntu 18.04
> 2. توصيل USB بجهاز تسلسلي
> 3. شاهده يتم تثبيته ضمن أجهزة COM في مدير الأجهزة
> 4. حاول فتح الجهاز في WSL2
> 5. القط /dev/ttyS[X]
> 
> * ما هو الخطأ / ما الذي يجب أن يحدث بدلاً من ذلك: 
>   يعمل هذا دون أي عيوب في WSL1، ولم يعد يعمل في WSL2 
>   في WSL2 يمكننا أن نرى بوضوح أنه غير متصل أو غير مهيأ 
>   /dev/ttyS3، UART: غير معروف، المنفذ: 0x02e8، IRQ: 3
>   ![COM3-wsl-pic](https://user-images.githubusercontent.com/1725627/61539227-6fe5ae00-aa11-11e9-87c8-aaff4fabf2ed.png)
> 
> ![COM3-pic](https://user-images.githubusercontent.com/1725627/61539193-5a708400-aa11-11e9-8331-9640b5ce35bb.png)

Osama

**Comment by h8920981945** [comment]
> It seems that we have a solution for the issue of @crose72 (and mine): https://github.com/dorssel/usbipd-win/issues/948#issuecomment-2138336528

**Comment by chri873** [comment]
> * Your Windows build number:  Microsoft Windows [Version 10.0.18932.1000]
> 
> * What you're doing and what's happening:  
> 1. run WSL2 Ubuntu 18.04
> 2. plugin USB to serial device
> 3. watch it mount under COM devices in device manager
> 4. attempt to open the device in WSL2
> 5. cat /dev/ttyS[X]
> 
> * What's wrong / what should be happening instead:
> This works under WSL1 flawlessy, it no longer works in WSL2
> in WSL2 we can clearly see it is not connected or configured
> /dev/ttyS3, UART: unknown, Port: 0x02e8, IRQ: 3
> ![COM3-wsl-pic](https://user-images.githubusercontent.com/1725627/61539227-6fe5ae00-aa11-11e9-87c8-aaff4fabf2ed.png)
> 
> ![COM3-pic](https://user-images.githubusercontent.com/1725627/61539193-5a708400-aa11-11e9-8331-9640b5ce35bb.png)
>
